src/core/cluster.py

```diff
@@ -35,9 +37,10 @@
 class ClusterState(Object):
     """Collection of global cluster state for the Kafka services."""
 
-    def __init__(self, charm: Framework | Object, substrate: Substrates):
+    def __init__(self, charm: Framework | Object, substrate: Substrates, config: CharmConfig):
         super().__init__(parent=charm, key="charm_state")
         self.substrate: Substrates = substrate
+        self.config = config
```

Comment:
**question**: Why not just get config through charm? With the right type it works the same:
```python
class ClusterState(Object):
    def __init__(self, charm: "KafkaCharm", substrate: Substrates):
        charm.config.<this is resolved to CharmConfig values>
```

Reply:
Good point! Have done.